### DISCUSSION TITLE
Overwrite a return type

### DISCUSSION THREAD

**Discussion by elbaro** [microsoft/pyright#7488] [Q&A] [answered]
I have a deroator that converts a generator to a normal function.
Since it's frequently used, I would like to omit `Generator` in the return type when the decorator is applied.

```py
@my_deco
def f() -> int:   #  -> Generator[..]
    yield 100
    yield 200
    return 300

assert f() == 300  # print(100), print(200), then return 300
```

Can Pyright be configured to check the return type after applying all decorators?

**Comment by Azureblade3808** [comment]
Firstly, try to post with a self-contained example next time. It's hard to understand what you're trying to do with both `my_deco` and all generic types in `Generator[...]` missing.

The short answer is NO. The return type `Generator[...]` can never be replaced with `int` in your code.

If you are asking for less typing, you can just omit the whole return type and let Pyright infer it for you. Or if you insist on specifying the generator return type, you may use an alias for shorter code --

```python
from typing import Any, Generator, Never

type G[RT] = Generator[Any, Never, RT]

def f() -> G[int]:
    yield 100
    yield 200
    return 300
```

**Reply by elbaro** [reply]
I am working on a large codebase that is already annotated with `-> RT`, so it's not an option.

mypy-plugin allows you to overwrite the return type. I know pyright does not have one yet, asked in a case there is a workaround for the decorator case.

**Reply by erictraut** [reply]
Pyright is a [standards-compliant type checker](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html). Since `int` is not a valid return type for a generator function, pyright is correct to generate a type error here. To fix this problem, you need to provide a valid return type annotation or delete the return type annotation and allow pyright to infer it for you. Providing a valid return type is preferable if you plan to use both pyright and mypy on your code base because mypy doesn't support return type inference.

You should assume that there will never be a plugin model for pyright. There are too many problems with plugin models, so we've ruled out that option. Our strategy is to continue to extend the python type system to support more behaviors. But that means you need to be compliant with the type system if you want to use pyright.

**Comment by rayansostenes** [comment, marked as answer]
It's should be easy to correctly type this decorator.
```python
from collections.abc import Callable, Generator
from typing import reveal_type


def my_deco[T, **P](func: Callable[P, Generator[int, None, T]], /) -> Callable[P, T]: ...


@my_deco
def f(a: str, b: int) -> Generator[int, None, int]:
    yield 100
    yield 200
    return 300


@my_deco
def g(a: str, b: int):
    yield 100
    yield 200
    return 300


reveal_type(f)  # Type of "f" is "(a: str, b: int) -> int"
reveal_type(g)  # Type of "g" is "(a: str, b: int) -> Literal[300]"
```
On pyright:
```console
$ npx pyright test.py
/tmp/test.py
  /tmp/test.py:22:13 - information: Type of "f" is "(a: str, b: int) -> int"
  /tmp/test.py:23:13 - information: Type of "g" is "(a: str, b: int) -> Literal[300]"
0 errors, 0 warnings, 2 informations
```

On mypy
> using `TypeVar` and `ParamSpec` since mypy does not support PEP 695 yet
```python
from collections.abc import Callable, Generator
from typing import Any, ParamSpec, TypeVar, reveal_type


T = TypeVar("T")
P = ParamSpec("P")


def my_deco(func: Callable[P, Generator[int, None, T]], /) -> Callable[P, T]: ...


@my_deco
def f(a: str, b: int) -> Generator[int, Any, int]:
    yield 100
    yield 200
    return 300


@my_deco
def g(a: str, b: int):
    yield 100
    yield 200
    return 300


reveal_type(f)  # Revealed type is "def (a: builtins.str, b: builtins.int) -> builtins.int"
reveal_type(g)  # Revealed type is "def (a: builtins.str, b: builtins.int) -> Any
```
```console
$ mypy test.py --allow-empty-bodies --python-version 3.12
test.py:26: note: Revealed type is "def (a: builtins.str, b: builtins.int) -> builtins.int"
test.py:27: note: Revealed type is "def (a: builtins.str, b: builtins.int) -> Any"
Success: no issues found in 1 source file
```
